### DISCUSSION TITLE
Logging Actions with Laravel Morph Maps

### DISCUSSION THREAD

**Discussion by crynobone** [laravel/nova-issues#3626] [Tips & Tricks]
Laravel Nova will store all actions completed and will utilise `Relation::morphMap()` if defined. It's also true with `Relation::enforceMorphMap()` however you have to be aware that you would need to define all Models and Pivot Models used to define Resource. 

As an example, imagine you have `Role` and `User` with belongs-to-many relationship and `Post`, `Tag` with morph-to-many relationship:

```php
<?php 

namespace App\Models;

use Illuminate\Foundation\Auth\User as Authenticatable;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\Pivot;
use Illuminate\Database\Eloquent\Relations\MorphPivot;

class User extends Authenticatable
{
    public function roles()
    {
        return $this->belongsToMany(Role::class)->using(UserRole::class);
    }
}

class Role extends Model
{
    public function users()
    {
        return $this->belongsToMany(User::class)->using(UserRole::class);
    }
}

class UserRole extends Pivot 
{
    //
}

class Tag extends Model
{
    public function posts()
    {
        return $this->morphToMany(Post::class, 'taggable')->using(Taggable::class);
    }
}

class Post extends Model
{
    public function tags()
    {
        return $this->morphToMany(Tag::class, 'taggable')->using(Taggable::class);
    }
}

class Taggable extends MorphPivot
{
    //
}

Relation::enforceMorphMap([
    'roles' => Role::class,
    'users' => User::class,
    'user_role' => UserRole::class,

    'posts' => Post::class,
    'tags' => Tag::class,
    'taggable' => Taggable::class,
]);
```

As shown above, you need to also define the Pivot Model used to allow Laravel Nova to properly log attaching and detaching roles events in the application when you used `Relation::enforceMorphMap()`.

**Comment by wivaku** [comment]
Is this a workaround or by design?

How is that used for polymorphic many to many (e.g. Tags from [Laravel](https://laravel.com/docs/8.x/eloquent-relationships#many-to-many-polymorphic-relations) and [Laravel Nova](https://nova.laravel.com/docs/1.0/resources/relationships.html#morphtomany) example)?

For each of the models using Tags I currently have:

```
public function tags()
{
   return $this->morphToMany(Tag::class, 'taggable');
}
```

[Laravel documentation](https://laravel.com/docs/8.x/eloquent-relationships#defining-custom-intermediate-table-models) states that:

> custom polymorphic many-to-many pivot models should extend the `Illuminate\Database\Eloquent\Relations\MorphPivot` class

Don't think the Tags example is a custom pivot model, but regardless: what would a complete pivot class example using `->using(...)` be for that case for both the `tags()` function and the `Relation::enforceMorphMap`?

**Reply by crynobone** [reply]
> Is this a workaround or by design?

It is by design if you want to enforce yourselves with strict using `enforceMorphMap()`, otherwise `morphMap()` would do fine. Most Laravel app doesn't interact directly on the belongs-to-many or morph-to-many on polymorphic relationship but when you do, you need to do this too.

> Don't think the Tags example is a custom pivot model, but regardless: what would a complete pivot class example using ->using(...) be for that case for both the tags() function and the Relation::enforceMorphMap?

Example updated using `Illuminate\Database\Eloquent\Relations\MorphPivot`

`MorphToMany` relationship extends `BelongsToMany`, so it'll be the same: https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Eloquent/Relations/MorphToMany.php
